src/main/java/seedu/duke/model/template/BooleanTemplate.java

```diff
@@ -138,6 +146,31 @@ private String getInputEquation(int index) throws DukeException {
         return equation;
     }
 
+    /**
+     * Recursive function which calculates the output value of the circuit at position passed.
+     *
+     * @param idx int type index of node.
+     * @return int type value output, can be 0 or 1.
+     * @throws DukeException If input values are not set.
+     */
+    public int calculateOutput(int idx) throws DukeException {
+        if (circuit.isNullAtIndex(idx)) {
+            throw new DukeException("Oops! Nothing set yet.");
+        }
+
+        boolean isNullAtRight = circuit.isNullAtIndex(circuit.getRightIndex(idx));
+        boolean isNullAtLeft = circuit.isNullAtIndex(circuit.getLeftIndex(idx));
+
+        if (!isNullAtLeft) {
+            circuit.getT(idx).setInput(calculateOutput(circuit.getLeftIndex(idx)));
```

Comment:
Consider initializing the `Gate` object before these `if` checks? Seems to be `circuit.getT(idx)` is accessed up to 3 times.

Reply:
I see, however since this is recursive, the Gate object will continuously be created again and again right...not sure if that's entirely a bad thing

Reply:
Okay I'll change and see

Reply:
No new `Gate` object is being created, you are accessing an existing one within `BinaryTree`. Plus you get much more clarity.

Reply:
On that note, you could also have a variable for the left and right index.